go/database/mpt/io/parallel_visit.go

```diff
@@ -0,0 +1,429 @@
+		for {
+			found := false
+			res, found = responses[cur]
+			if found {
+				delete(responses, cur)
+				break
+			}
+			responsesCond.Wait()
+		}
+		responsesMutex.Unlock()
+
+		if res.err != nil {
+			return res.err
+		}
+
+		switch visitor.Visit(res.node, mpt.NodeInfo{Id: cur}) {
+		case mpt.VisitResponseAbort:
+			return nil
+		case mpt.VisitResponsePrune:
+			continue
+		}
+
+		switch node := res.node.(type) {
+		case *mpt.BranchNode:
+			children := node.GetChildren()
+			for i := len(children) - 1; i >= 0; i-- {
+				id := children[i].Id()
+				if !id.IsEmpty() {
+					stack = append(stack, id)
+				}
+			}
+		case *mpt.ExtensionNode:
+			next := node.GetNext()
+			stack = append(stack, next.Id())
+		case *mpt.AccountNode:
+			if !cutAtAccounts {
+				storage := node.GetStorage()
+				id := storage.Id()
+				if !id.IsEmpty() {
+					stack = append(stack, id)
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+type position struct {
+	parent *position
+	pos    byte
+	len    byte
+}
+
+func newPosition(pos []byte) *position {
+	var res *position
+	for i, step := range pos {
+		res = &position{
+			parent: res,
+			pos:    step,
+			len:    byte(i),
+		}
+	}
+	return res
+}
+
+func (p *position) String() string {
+	if p == nil {
+		return ""
+	}
+	if p.parent == nil {
+		return fmt.Sprintf("%d", p.pos)
+	}
+	return fmt.Sprintf("%s.%d", p.parent.String(), p.pos)
+}
+
+func (p *position) Child(step byte) *position {
+	len := byte(0)
+	if p != nil {
+		len = p.len
+	}
+	return &position{
+		parent: p,
+		pos:    step,
+		len:    len + 1,
+	}
+}
+
+func (p *position) compare(b *position) int {
+	if p == b {
+		return 0
+	}
+	if p == nil {
+		return -1
+	}
+	if b == nil {
+		return 1
+	}
+
+	// make sure a is the shorter one
+	if p.len > b.len {
+		return b.compare(p) * -1
+	}
+
+	// reduce the length of b to match a
+	bIsLonger := p.len < b.len
+	for p.len < b.len {
+		b = b.parent
+	}
+
+	// compare the common part
+	prefixResult := p._compare(b)
+	if prefixResult != 0 {
+		return prefixResult
+	}
+	if bIsLonger {
+		return -1
+	}
+	return 0
+}
+
+func (p *position) _compare(b *position) int {
+	if p == b {
+		return 0
+	}
+	prefixResult := p.parent._compare(b.parent)
+	if prefixResult != 0 {
+		return prefixResult
+	}
+	if p.pos < b.pos {
+		return -1
+	}
+	if p.pos > b.pos {
+		return 1
+	}
+	return 0
+}
+
+// ----------------------------------------------------------------------------
+//                               NodeSource
+// ----------------------------------------------------------------------------
+
+type nodeSourceFactory struct {
+	directory string
+}
+
+func (f *nodeSourceFactory) Open() (NodeSource, error) {
+	accounts, err := openSource[mpt.AccountNode](f.directory, "accounts", mpt.AccountNodeWithPathLengthEncoderWithNodeHash{})
+	if err != nil {
+		return nil, err
+	}
+	branches, err := openSource[mpt.BranchNode](f.directory, "branches", mpt.BranchNodeEncoderWithNodeHash{})
+	if err != nil {
+		return nil, err
```

Comment:
If opening the `branches` failed, `accounts` should be closed. The same for the sources below.

Reply:
good point, fixed